src/bin/pg_autoctl/cli_perform.c

```diff
@@ -304,3 +316,68 @@ cli_perform_failover(int argc, char **argv)
 		exit(EXIT_CODE_INTERNAL_ERROR);
 	}
 }
+
+
+/*
+ * cli_perform_promotion calls the function pgautofailover.perform_promotion()
+ * on the monitor.
+ */
+static void
+cli_perform_promotion(int argc, char **argv)
+{
+	Keeper keeper = { 0 };
+	Monitor *monitor = &(keeper.monitor);
+	KeeperConfig *config = &(keeper.config);
+
+	int groupId = 0;
+
+	PgInstanceKind nodeKind = NODE_KIND_UNKNOWN;
+
+	char *channels[] = { "state", NULL };
+
+	keeper.config = keeperOptions;
+
+	if (!monitor_init_from_pgsetup(monitor, &keeper.config.pgSetup))
+	{
+		/* errors have already been logged */
+		exit(EXIT_CODE_BAD_ARGS);
+	}
+
+	/* grab --name from either the command options or the configuration file */
+	(void) cli_ensure_node_name(&keeper);
+
+	if (!monitor_get_groupId_from_name(monitor,
+									   config->formation,
+									   config->name,
+									   &groupId))
+	{
+		/* errors have already been logged */
+		exit(EXIT_CODE_BAD_ARGS);
+	}
+
+	/* start listening to the state changes before we call perform_promotion */
+	if (!pgsql_listen(&(monitor->pgsql), channels))
+	{
+		log_error("Failed to listen to state changes from the monitor");
+		exit(EXIT_CODE_MONITOR);
+	}
+
+	/*
+	 * pgautofailover.perform_promotion returns true when a promotion has been
+	 * triggered, and false when none was necessary. When an error occurs, it
+	 * reports an error condition, which is logged about already.
+	 */
+	if (monitor_perform_promotion(monitor, config->formation, config->name))
+	{
+		/* process state changes notification until we have a new primary */
+		if (!monitor_wait_until_some_node_reported_state(monitor,
```

Comment:
As far as I can see, `monitor_wait_until_some_node_reported_state ` has  a 60 seconds timeout, after that we see 
```
12:43:31 65553 ERROR Failed to receive monitor's notifications
12:43:31 65553 ERROR Failed to wait until a new primary has been notified
```

We could likely to exceed the timeout on some real-world scenarios with async replicas. 

If my above observation is accurate, I can see two options: (a) wait indefinitely -- which sounds scary :)   (b) somehow check that LSN is increasing (such  as `standby_fetch_missing_wal` )and reset the timeout as long as the LSN increases

Reply:
If there's no activity (downtime and maintenance windows) then the LSN will not increase either. I think it's fine to have a timeout here, and then users can resort to `watch -n 1 pg_autoctl show state` or their favorite way to follow along what's happening. It might also be a good time to have a look at the logs to understand what's going on, in the case when it's not just taking more time...

Reply:
>  I think it's fine to have a timeout here

With the current code, as far as I can see, the promotion would fail even if everything is OK, only if just it takes some time to fetch the wal (in case of async replication) (or some checkpointing?). Am I missing anything on this?

I think we should definitely have the timeout, but it should not break a potentially viable promotion.

> If there's no activity (downtime and maintenance windows) then the LSN will not increase either.

Yeah, in that case we should be fine to fail the promotion anyway?

I think in the vast majority of the cases we would not hit this timeout, but still I can see that some users might hit this. The workaround is to set the priority to highest value, and execute `perform switchover`. 

If we are going with the current code -- which I'm fine with --, it could be nice to document it somewhere (either a github issue or the documentation).

Reply:
I'm confused now. The timeout is applied to the client part where we wait for notifications. The only impact when we reach the timeout is that we stop displaying the state changes that are happening on the monitor. The state changes are still happening. It might even be a network timeout problem, after all. The promotion/failover is happening, we just are not receiving notifications about it. Or it's taking a long time, and it is continuing in the background.

Reaching this timeout has no impact on the failover mechanics.